virtio-devices/src/mem.rs

```diff
@@ -758,8 +828,59 @@ impl Mem {
             config: Arc::new(Mutex::new(config)),
             seccomp_action,
             hugepages,
+            dma_mapping_handlers: Arc::new(Mutex::new(BTreeMap::new())),
+            blocks_state: Arc::new(Mutex::new(BlocksState(vec![
+                false;
+                (config.region_size / config.block_size)
+                    as usize
+            ]))),
         })
     }
+
+    pub fn add_dma_mapping_handler(
+        &mut self,
+        device_id: u32,
+        handler: Arc<dyn ExternalDmaMapping>,
+    ) -> result::Result<(), Error> {
+        let config = self.config.lock().unwrap();
+
+        if config.plugged_size > 0 {
+            for (idx, plugged) in self.blocks_state.lock().unwrap().inner().iter().enumerate() {
+                if *plugged {
+                    let gpa = config.addr + (idx as u64 * config.block_size);
+                    handler
+                        .map(gpa, gpa, config.block_size)
+                        .map_err(Error::DmaMap)?;
+                }
+            }
+        }
+
+        self.dma_mapping_handlers
+            .lock()
+            .unwrap()
+            .insert(device_id, handler);
+
+        Ok(())
+    }
+
+    pub fn remove_dma_mapping_handler(&mut self, device_id: u32) -> result::Result<(), Error> {
+        if let Some(handler) = self.dma_mapping_handlers.lock().unwrap().remove(&device_id) {
+            let config = self.config.lock().unwrap();
+
+            if config.plugged_size > 0 {
+                for (idx, plugged) in self.blocks_state.lock().unwrap().inner().iter().enumerate() {
+                    if *plugged {
+                        let gpa = config.addr + (idx as u64 * config.block_size);
+                        handler
+                            .unmap(gpa, config.block_size)
+                            .map_err(Error::DmaUnmap)?;
+                    }
+                }
+            }
+        }
```

Comment:
Should there be an error logged if you try and remove for an unknown device?

You could do something like this `let handler = self.dma_mapping_handlers.lock().unwrap().remove(&device_id).ok_or(Error::InvalidDeviceHandler)?;`

Reply:
Good point!